[Refactor/Bug/Move] Overhaul Stats and Battle Items, Implement Several Stat Moves

xsn34kzx opened the pull request:
<!-- Make sure the title includes categorization (i.e. [Bug], [QoL], [Localization]) -->
<!-- Make sure that this PR is not overlapping with someone else's work -->
<!-- Please try to keep the PR self-contained (and small) -->

## What are the user-related changes?
<!-- Summarize what are the changes from a user perspective on the application -->

- "Natural" stat-related fixes to moves, abilities, and other miscellaneous things:
  - [`Beast Boost`](https://bulbapedia.bulbagarden.net/wiki/Beast_Boost_(Ability)) now considers all effective stats when choosing a stat stage to boost, considers temporarily overridden stats due to moves like `Speed Swap`, and honors order preference when choosing a stat stage to boost (based on the highest stat).
  - `Imposter` and `Transform` now have their copied stats reflected in effective stat calculations (related to #2564 but does not fully address it). This has effects in enemy AI, how damage is calculated, and much more.
  - `White Smoke`/`Clear Body` no longer push multiple text prompts announcing their effects for instances where multiple stat stage decreases occur (i.e., getting hit by `Tickle` or `Parting Shot`).
  - When a transformed Pokemon leveled, the level-up phase would display and use the copied stats as the basis for calculating new stats; now, the Pokemon's true stats are always reflected.
  - When a boss segment is cleared, all effective stats are now weighted properly, which means that the chance of the boss getting the appropriate stat stage raises is more accurate.
- Implements several moves to essentially demo the overhaul:
  - [`Power Split`](https://bulbapedia.bulbagarden.net/wiki/Power_Split_(move))
  - [`Guard Split`](https://bulbapedia.bulbagarden.net/wiki/Guard_Split_(move))
  - [`Speed Swap`](https://bulbapedia.bulbagarden.net/wiki/Speed_Swap_(move))
  - [`Guard Swap`](https://bulbapedia.bulbagarden.net/wiki/Guard_Swap_(move))
  - [`Power Swap`](https://bulbapedia.bulbagarden.net/wiki/Power_Swap_(move))
- Overhauls the existing battle items (at request of the balance team):
  - Only one of a given type of item can be held at a time.
    - e.g., if you already have an `X Attack`, you can't have another `X Attack` but grabbing an `X Accuracy` happens without a hitch.
  - To accommodate the above change, grabbing another item of the same type will now "refresh" the battle count of the one currently obtained.
    - e.g., taking a new `X Defense` when you have an `X Defense` at 3 battles left will bring that existing `X Defense` back to 5 battles left but won't add the new one to the modifier list.
  - Every battle item now only provides a x1.3 boost instead of the original x1.5 boost with the exception of `X Accuracy` and `Dire Hit`, which remain unchanged.
    - e.g., with an `X Attack` and the `ATK` stat stage raised by 1, you'd get a x1.8 multiplier to ATK's effective stat value; note that this still caps at x4.

## What are the developer-related changes?

- Incorporates `summonData.stats` into a setter and getter with an `ignoreOverride` parameter, which follows the precedent set by getters like `getGender`, `getSpeciesForm`, etc.
  - All usages of `stats` and `summonData.stats` are replaced with this function to have better control of the behavior globally.
- `summonData.battleStats` has now be renamed to `summonData.statStages` along with a setter and getter dedicated to stat stages; this name change follows through to all attribute classes that alter stat stages but are labeled `StatChanges`, `BattleStat` or otherwise (see **Motivation and Rationale**).
  - Most usages of `summonData.battleStats` has been replaced for similar reasons above.
- All stat-related enums (`Stat`, `BattleStat`, and `TempBattleStat`) have been restructured such that `Stat` is the sole enum; however, to take advantage of ensuring that only certain `Stat` enum values are used in or at given point of the code base, new types have been created to provided guard rails on what `Stat` enum values can be passed in and used. All of these types are bundled with their own constant array as well (see **Motivation and Rationale**).
  - `PermanentStat` is a new type that refers to `HP`, `ATK`, `DEF`, `SPATK`, `SPDEF`, and `SPD`, which are the core stats of a Pokemon that exist in and out of battle.
  - `BattleStat` has been converted to a type that refers to `ATK`, `DEF`, `SPATK`, `SPDEF`, `SPD`, `ACC`, and `EVA`, which are stats that have applicable in-battle stat stages.
  - `EffectiveStat` is a new type that refers to `ATK`, `DEF`, `SPATK`, `SPDEF`, and `SPD`, which is the intersection between the core stats of a Pokemon and the stats that have relevance in battle.
  - `TempBattleStat` has been converted to a type that refers `ATK`, `DEF`, `SPATK`, `SPDEF`, `SPD`, and `ACC`. This more or less will only have usage as a pre-generator argument than anything.
- Refactors across the board for anything that touched stats, ranging from simple search-and-replacements to rewrites:
   - `TempBattleStatBooster`s (battle items) were renamed (`TempStatStageBooster`) and refactored, removing duplicate helper functions along the way. The generator class was made into a named class instead of an anonymous one.
   - `BaseStatBooster`s (vitamins) were refactored, removing duplicate helper functions along the way. The generator class was made into a named class instead of an anonymous one.
   - `PokemonResetStatStageModifier`, (`White Herb`), although not used at the moment, was renamed (`ResetStatStageModifier`) and refactored.
   - Minor miscellaneous optimizations for random things concerning stats and going through stats within the code base.
     - A good highlight of this was getting rid of every instance of `Utils.getEnumValues(Stat)` and using the new constant arrays.
   - A good chunk of unit tests had to be changed to accommodate the enum change; while doing this, I also took the chance to clean up tests along the way if applicable or get rid of superfluous tests.
     - A general rule of thumb I had was to keep anything that was unique, specific, or happening with the expectation different than a previously test scenario. For example, I took out some tests from `Intimidate` because it tested the ability in different scenarios that had no bearing on the actual ability. I didn't do this very often though, since the point of the PR isn't necessarily to do a full unit test once-over.
     - I also removed two entire test files that concerned `BattleStat`s specifically since the enum was removed and the functions it tested were reworked.
   - Bonus inconsequential but helpful changes
     - Extracted stat stage multiplier and critical hit stage calculation into their own functions for spying purposes in unit tests (examples for both can be seen in `temp_stat_stage_booster.test.ts` and `dire_hit.test.ts`).

## Motivation and Rationale
<!-- Explain why you decided to introduce these changes -->
<!-- Does it come from an issue or another PR? Please link it -->
<!-- Explain why you believe this can enhance user experience -->

A large chunk of the necessity for a PR to become this big comes more or less from how ingrained the existing stats logic and nomenclature has been embedded in the code base, which in general is confusing and inconsistent. With `BattleStat` referring to both stats used in battle and stat stages, many classes end up having names that aren't very descriptive and get proliferated on the pretense of stability or consistency, which is not the case. Trying to discern what stats are what ("what's the difference between `stats`, `summonData.stats`, `summonData.battleStats`, and `baseStats`?") requires more investigation than it should warrant and having one identifier mean three different things (`summonData.battleStats`, `BattleStat`, and `getBattleStat`) becomes increasingly frustrating when having to go into older portions of code for debugging, extending, or refactoring. 

This leads to the why behind breaking down the three existing stat-related enums into one: it becomes increasingly hard to discern which enum should apply, especially in cases in which you'd want to limit what stats can and can't be used in particular parts of the code base (that is, the differentiation between `Stat` and `BattleStat` *does* exist, but to prevent the current `getBattleStat` function from the user providing `Stat.HP` as an argument, we're forced to use an `if` to redirect to another function). Because they also already share a large number of values as is, using either interchangeably can go by without being noticed or scrutinized to deeply when those stats should not apply in the scenario (for example, you shouldn't be "forced" to use `BattleStat` for scenarios where you want the effective stat of a Pokemon, leaving yourself open to future contributors potentially providing `Stat.ACC` or `Stat.EVA` and needing to write a catch for those). Besides the issue of legibility, it, more than anything, encourages manual casting manipulating the stat enums as values in order to jump between the two classifications. By coalescing them into one, we get the benefit of providing error feedback when using stat values that shouldn't be used in that scenario and only ever having to worry about one enum across all scenarios. In regards to their specific implementation as types based on arrays of values rather than using an `Exclude` statement, this was primarily motivated by how often `Utils.getEnumValues(Stat)` was used throughout the code base, adding unnecessary complexity for what should be a global array; additionally, they give us the ability of have ergonomic, typed `for` loops (a great example of this is in `TransformAttr` in `src/data/move.ts`).

This seeps into the why behind overhauling battle items in this PR: by dissolving `TempBattleStat` and using `Stat` instead as the backing enum, all battle items will essentially "break", (see **Leftovers and Caveats**) so we might as well take the opportunity to change how they work fundamentally. My own goal here was to revamp very old code and break down single-use helper functions that shared overlap with other helper functions; the in-game changes come at the request of the balance team given the how strong having several battle items of the same type are (`X Speed` turning Pokemon into natural sweepers for five or more turns, etc.).

As an aside, the motivation for naming comes mostly from Bulbapedia's relatively consistent naming across its pages:
- [Stat Stages](https://bulbapedia.bulbagarden.net/wiki/Stat_modifier#Stages)
- [Permanent Stat](https://bulbapedia.bulbagarden.net/wiki/Stat#Permanent_stats)
- [Effective Stat](https://bulbapedia.bulbagarden.net/wiki/Stat#Stat_modifiers)
- [Battle Stat](https://bulbapedia.bulbagarden.net/wiki/Stat#In-battle_stats)

Matching the naming conventions here in itself has value when implementing new moves, abilities, etc., making it easier to translate effects just by reading rather than automatically being opted into having to do more research for more clarity.

Finally, the implemented moves come bundled to mostly demonstrate the leg work of overhaul (as particularly requested by @f-fsantos), taking advantage of linking `getStats` to both `stats` and `summonData.stats` and pivoting between the two when necessary. They were relatively simple implementations, so it gives a chance to stress test and begin the discussion on their particular move interactions outside of what they should already do by default.

## Leftovers and Caveats

There are a few caveats that come out of a PR this big:
- I tried to add unit tests to all the new things I did to double check the work that was done along with making sure that all existing unit tests still passed. A couple Classic Mode runs were played on my end, but I wouldn't doubt if there are potentially undocumented anomalies, so I'd like more manual testing to be done in general.
- With the complete overhaul of the battle items, there is a slight issue when going to production: all existing battle items will get "flushed" out of any sessions that contain them. This doesn't break saves as far as I know, but it does remove them from those sessions on reload since the game can't find their new name in outdated session data. Moreover, even if the entry name was changed back to its original name, the pre-generator argument would not corresponding to the original `TempBattleStat` enum values (in that enum, `0` -> `ATK`, which differs from Stat where `1` -> `ATK`). As such, existing battle items are "broken" either way if there is no save conversion module. This might not be a bad trade off since they're expiring modifiers anyway, but that's not necessarily my final decision.
- Currently, moves like `Transform`, `Speed Swap`, etc. operate on the pretense that they copy overridden stats; that is, if I use `Transform` on a Pokemon who used `Speed Swap` with a previous Pokemon, the transformed Pokemon would get that the stat resulting from `Speed Swap` instead of that Pokemon's original stat. There's not a lot of specific documentation on the interactions between those moves, but, for now, I operated on the logic.
- Lastly, it's still being debated whether to include `Power Trick`. It's a tad more complex than any of the moves included already (and has an existing draft PR), but I'm willing to take that on after this PR if it needs a mirror.

### Screenshots/Videos
<!-- If your change is changing anything on the user experience, please provide visual proofs of it -->
<!-- Please take screenshots/videos before and after your changes, to show what is brought by this PR -->

#### Moves

| Power Swap | Guard Swap |
| --- | --- |
| <img width="365" alt="powerSwap" src="https://github.com/user-attachments/assets/39e082b5-b4cc-45bf-99ae-5f341ec9b9cf"> | <img width="365" alt="guardSwap" src="https://github.com/user-attachments/assets/560d1b60-414b-4dbd-b662-b2bd70a19701"> |

| Power Split | Guard Split |
| --- | --- |
| <img width="365" alt="powerSplit" src="https://github.com/user-attachments/assets/c2ca990e-129e-4002-9ffe-eb1314cd11d6"> | <img width="365" alt="guardSplit" src="https://github.com/user-attachments/assets/29418c3d-e040-423f-b727-d4e6b04fd8ed"> |

| Speed Swap |
| --- |
| <img width="365" alt="speedSwap" src="https://github.com/user-attachments/assets/75415e58-89af-4081-b62d-71e1fe547661"> |

#### Battle Items

##### Before
https://github.com/user-attachments/assets/278499e8-d62f-4bf4-8615-056be14bbdb8

##### After
https://github.com/user-attachments/assets/eb316e5f-42fd-4fad-a0b8-63b03362848d

## How to test the changes?
<!-- How can a reviewer test your changes once they check out on your branch? -->
<!-- Did you just make use of the `src/overrides.ts` file? -->
<!-- Did you introduce any automated tests? -->
<!-- Do the reviewer need to do something special in order to test your change? -->

The bundled unit tests essentially cover the core and niche scenarios that I aimed to address most specifically, but I'd encourage playing the game in general to make sure no new bugs arise from this PR more than anything.

## Checklist
- [x] Does this overlap or block another PR?
  - #2658
  - #1035 
  - #3394
- [ ] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

torranx commented:
would love to see this implemented. lmk if i can help with unit tests, feel free to @ me on discord (Swain)

xsn34kzx commented:
> would love to see this implemented. lmk if i can help with unit tests, feel free to @ me on discord (Swain)

@torranx Definitely a ways away for now, but I'll definitely let you know when everything is mostly squared away 👍🏼

ysh4296 commented:
It seems this pr has a wider scope of utility.

its good idea to use `summonData.stats` for Override stats

i'll set #2658 to draft and delete it after this pr merged.

xsn34kzx commented:
> It seems this pr has a wider scope of utility.
> 
> its good idea to use `summonData.stats` for Override stats
> 
> i'll set #2658 to draft and delete it after this pr merged.

@ysh4296 I was actually thinking about it more, and I think it'd be best if it was possible to use a lot of the work you already did instead of completely getting rid of it (especially since there's a lot of translation work done and I think the brain-stormed idea I had was similar to your use of battle tags). I could probably leave out `Power Trick` and focus on the other moves for now since I hadn't really started it yet, and then you could probably take the lead on it again after this gets merged to avoid losing the progress and probably streamline implementation since there's really no easy way to do it without fixing `summonData.stats` first.

ysh4296 commented:
> @ysh4296 I was actually thinking about it more, and I think it'd be best if it was possible to use a lot of the work you already did instead of completely getting rid of it (especially since there's a lot of translation work done and I think the brain-stormed idea I had was similar to your use of battle tags). I could probably leave out `Power Trick` and focus on the other moves for now since I hadn't really started it yet, and then you could probably take the lead on it again after this gets merged to avoid losing the progress and probably streamline implementation since there's really no easy way to do it without fixing `summonData.stats` first.

@xsn34kzx 
thank you. 
I agree that this work should be done before #2658
I will keep an eye on this PR

xsn34kzx commented on src/test/items/eviolite.test.ts:
```diff
@@ -37,29 +37,29 @@ describe("Items - Eviolite", () => {
 
-    // Checking consoe log to make sure Eviolite is applied when getBattleStat (with the appropriate stat) is called
-    partyMember.getBattleStat(Stat.DEF);
+    // Checking consoe log to make sure Eviolite is applied when getEffectiveStat (with the appropriate stat) is called
```
```suggestion
    // Checking console log to make sure Eviolite is applied when getEffectiveStat (with the appropriate stat) is called
```
NIT: Minor correction

xsn34kzx commented on src/test/items/light_ball.test.ts:
```diff
@@ -37,29 +37,29 @@ describe("Items - Light Ball", () => {
 
-    // Checking consoe log to make sure Light Ball is applied when getBattleStat (with the appropriate stat) is called
-    partyMember.getBattleStat(Stat.DEF);
+    // Checking consoe log to make sure Light Ball is applied when getEffectiveStat (with the appropriate stat) is called
```
```suggestion
    // Checking console log to make sure Light Ball is applied when getEffectiveStat (with the appropriate stat) is called
```
NIT: Minor correction

xsn34kzx commented on src/test/items/metal_powder.test.ts:
```diff
@@ -37,29 +37,29 @@ describe("Items - Metal Powder", () => {
 
-    // Checking consoe log to make sure Metal Powder is applied when getBattleStat (with the appropriate stat) is called
-    partyMember.getBattleStat(Stat.DEF);
+    // Checking consoe log to make sure Metal Powder is applied when getEffectiveStat (with the appropriate stat) is called
```
```suggestion
    // Checking console log to make sure Metal Powder is applied when getEffectiveStat (with the appropriate stat) is called
```
NIT: Minor correction

xsn34kzx commented on src/test/items/quick_powder.test.ts:
```diff
@@ -37,29 +37,29 @@ describe("Items - Quick Powder", () => {
 
-    // Checking consoe log to make sure Quick Powder is applied when getBattleStat (with the appropriate stat) is called
-    partyMember.getBattleStat(Stat.DEF);
+    // Checking consoe log to make sure Quick Powder is applied when getEffectiveStat (with the appropriate stat) is called
```
```suggestion
    // Checking console log to make sure Quick Powder is applied when getEffectiveStat (with the appropriate stat) is called
```
NIT: Minor correction

xsn34kzx commented on src/test/moves/baton_pass.test.ts:
```diff
@@ -45,24 +45,27 @@ describe("Moves - Baton Pass", () => {
     game.override
       .startingWave(5)
-      .enemyMoveset(new Array(4).fill([Moves.NASTY_PLOT]));
+      .enemyMoveset([ Moves.NASTY_PLOT, Moves.NASTY_PLOT, Moves.NASTY_PLOT, Moves.NASTY_PLOT ]);
```
```suggestion
      .enemyMoveset(new Array(4).fill([Moves.NASTY_PLOT]);
```
NIT: Reverting change

xsn34kzx commented on src/test/moves/baton_pass.test.ts:
```diff
@@ -74,13 +77,13 @@ describe("Moves - Baton Pass", () => {
     // round 2 - baton pass
     game.scene.getEnemyPokemon()!.hp = 100;
-    game.override.enemyMoveset(new Array(4).fill(Moves.BATON_PASS));
+    game.override.enemyMoveset([ Moves.BATON_PASS, Moves.BATON_PASS, Moves.BATON_PASS, Moves.BATON_PASS ]);
```
```suggestion
    game.override.enemyMoveset(new Array(4).fill(Moves.BATON_PASS));
```
NIT: Reverting change

xsn34kzx commented on src/test/moves/spit_up.test.ts:
```diff
@@ -34,8 +36,10 @@ describe("Moves - Spit Up", () => {
     game.override.enemyLevel(2000);
 
-    game.override.moveset([Moves.SPIT_UP, Moves.SPIT_UP, Moves.SPIT_UP, Moves.SPIT_UP]);
+    game.override.moveset([ spitUp.id, spitUp.id, spitUp.id, spitUp.id ]);
```
```suggestion
    game.override.moveset(new Array(4).fill(spitUp.id));
```
NIT: Match previous NIT

xsn34kzx commented on src/ui/battle-message-ui-handler.ts:
```diff
@@ -3,11 +3,11 @@ import { addBBCodeTextObject, addTextObject, getTextColor, TextStyle } from "./t
 import BBCodeText from "phaser3-rex-plugins/plugins/bbcodetext";
 import {Button} from "#enums/buttons";
 import i18next from "i18next";
+import { Stat, getStatKey, PERMANENT_STATS } from "#app/enums/stat.js";
```
```suggestion
import { Stat, PERMANENT_STATS, getStatKey } from "#app/enums/stat";
```
NIT: Label types and standardize order

xsn34kzx replied:
```suggestion
import { Stat, PERMANENT_STATS, getStatKey } from "#app/enums/stat";
```

xsn34kzx commented on src/ui/stats-container.ts:
```diff
@@ -1,7 +1,8 @@
 import BattleScene from "../battle-scene";
-import { Stat, getStatName } from "../data/pokemon-stat";
 import { TextStyle, addBBCodeTextObject, addTextObject, getTextColor } from "./text";
+import { getStatKey, PERMANENT_STATS } from "#app/enums/stat.js";
```
```suggestion
import { PERMANENT_STATS, getStatKey } from "#app/enums/stat.js";
```
NIT: Label types and standardize order

xsn34kzx commented on src/ui/summary-ui-handler.ts:
```diff
@@ -23,6 +22,7 @@ import { Ability } from "../data/ability.js";
 import i18next from "i18next";
 import {modifierSortFunc} from "../modifier/modifier";
 import { PlayerGender } from "#enums/player-gender";
+import { Stat, getStatKey, PERMANENT_STATS } from "#app/enums/stat.js";
```
```suggestion
import { Stat, PERMANENT_STATS, getStatKey } from "#app/enums/stat.js";
```
NIT: Label types and standardize order

xsn34kzx commented on src/phases.ts:
```diff
@@ -66,6 +64,7 @@ import { Species } from "#enums/species";
 import { TrainerType } from "#enums/trainer-type";
 import { applyChallenges, ChallengeType } from "./data/challenge";
 import { pokemonEvolutions } from "./data/pokemon-evolutions";
+import { Stat, BattleStat, getStatKey, getStatStageChangeDescriptionKey } from "#enums/stat";
```
```suggestion
import { Stat, type BattleStat, getStatKey, getStatStageChangeDescriptionKey } from "#enums/stat";
```
NIT: Label types and standardize order

xsn34kzx commented on src/field/pokemon.ts:
```diff
@@ -51,6 +49,7 @@ import { Biome } from "#enums/biome";
 import { Moves } from "#enums/moves";
 import { Species } from "#enums/species";
 import { getPokemonNameWithAffix } from "#app/messages.js";
+import { PermanentStat, BattleStat, PERMANENT_STATS, EffectiveStat } from "#app/enums/stat"; // TODO: Add type
```
```suggestion
```
NIT: Remove in favor of one import line

xsn34kzx commented on src/field/pokemon.ts:
```diff
@@ -3,27 +3,25 @@ import BattleScene, { AnySound } from "../battle-scene";
 import { getLevelTotalExp } from "../data/exp";
-import { Stat } from "../data/pokemon-stat";
-import { DamageMoneyRewardModifier, EnemyDamageBoosterModifier, EnemyDamageReducerModifier, EnemyEndureChanceModifier, EnemyFusionChanceModifier, HiddenAbilityRateBoosterModifier, PokemonBaseStatModifier, PokemonFriendshipBoosterModifier, PokemonHeldItemModifier, PokemonNatureWeightModifier, ShinyRateBoosterModifier, SurviveDamageModifier, TempBattleStatBoosterModifier, StatBoosterModifier, CritB ...
+import { BATTLE_STATS, EFFECTIVE_STATS, Stat } from "#enums/stat";
```
```suggestion
import { Stat, type PermanentStat, type BattleStat, type EffectiveStat, PERMANENT_STATS, BATTLE_STATS, EFFECTIVE_STATS } from "#enums/stat";
```
NIT: Label types and standardize order

xsn34kzx commented on src/data/ability.ts:
```diff
@@ -26,6 +24,7 @@ import { ArenaTagType } from "#enums/arena-tag-type";
 import { BattlerTagType } from "#enums/battler-tag-type";
 import { Moves } from "#enums/moves";
 import { Species } from "#enums/species";
+import { Stat, BattleStat, getStatKey, BATTLE_STATS, EFFECTIVE_STATS, EffectiveStat } from "#app/enums/stat.js";
```
```suggestion
import { Stat, type BattleStat, type EffectiveStat, BATTLE_STATS, EFFECTIVE_STATS, getStatKey } from "#app/enums/stat";
```
NIT: Label types and standardize order

xsn34kzx reviewed (commented):
Did a gloss over of everything on my end to catch some of the more obvious issues/oversights while I did everything.

xsn34kzx commented on src/ui/battle-message-ui-handler.ts:
```diff
@@ -1,13 +1,12 @@
 import BBCodeText from "phaser3-rex-plugins/plugins/bbcodetext";
 import {Button} from "#enums/buttons";
 import i18next from "i18next";
+import { Stat, PERMANENT_STATS, getStatKey } from "#app/enums/stat.js";
```
```suggestion
import { Stat, PERMANENT_STATS, getStatKey } from "#app/enums/stat";
```
NIT: Get rid of extension

xsn34kzx commented on src/test/items/thick_club.test.ts:
```diff
@@ -37,29 +37,29 @@ describe("Items - Thick Club", () => {
 
-    // Checking consoe log to make sure Thick Club is applied when getBattleStat (with the appropriate stat) is called
-    partyMember.getBattleStat(Stat.DEF);
+    // Checking consoe log to make sure Thick Club is applied when getEffectiveStat (with the appropriate stat) is called
```
```suggestion
    // Checking console log to make sure Thick Club is applied when getEffectiveStat (with the appropriate stat) is called
```
NIT: Minor correction

torranx commented on src/data/ability.ts:
```diff
@@ -1031,22 +1028,22 @@ export class PostDefendMoveDisableAbAttr extends PostDefendAbAttr {
-  applyPostStatChange(pokemon: Pokemon, statsChanged: BattleStat[], levelsChanged: integer, selfTarget: boolean, args: any[]): boolean {
-    if (this.condition(pokemon, statsChanged, levelsChanged) && !selfTarget) {
-      pokemon.scene.unshiftPhase(new StatChangePhase(pokemon.scene, (pokemon).getBattlerIndex(), true, this.statsToChange, this.levels));
+  applyPostStatStageChange(pokemon: Pokemon, statsChanged: BattleStat[], stagesChanged: integer, selfTarget: boolean, args: any[]): boolean {
```
nit
```suggestion
  applyPostStatStageChange(pokemon: Pokemon, statsChanged: BattleStat[], stagesChanged: number, selfTarget: boolean, args: any[]): boolean {
```

torranx commented on src/data/ability.ts:
```diff
@@ -2406,7 +2395,7 @@ export class ProtectStatAbAttr extends PreStatChangeAbAttr {
   }
 
-  applyPreStatChange(pokemon: Pokemon, passive: boolean, stat: BattleStat, cancelled: Utils.BooleanHolder, args: any[]): boolean {
+  applyPreStatStageChange(_pokemon: Pokemon, _passive: boolean, stat: BattleStat, cancelled: Utils.BooleanHolder, _args: any[]): boolean {
```
any reason on why the sudden param naming convention change?

xsn34kzx replied:
A more board motivation for it is in the main description, but the TL;DR is for better categorization of what topics correspond to what. Right now, we have `BattleStats` and `StatChanges` effectively pointing to the same concept of dealing with stat stage boosts or changes for example, so really the idea is the make it so, for things like searching, all you need is one key word and we all know what's being discussed.

torranx commented on src/data/ability.ts:
```diff
@@ -3412,9 +3400,9 @@ export class StatChangeMultiplierAbAttr extends AbAttr {
+export class StatStageChangeCopyAbAttr extends AbAttr {
   apply(pokemon: Pokemon, passive: boolean, cancelled: Utils.BooleanHolder, args: any[]): boolean | Promise<boolean> {
-    pokemon.scene.unshiftPhase(new StatChangePhase(pokemon.scene, pokemon.getBattlerIndex(), true, (args[0] as BattleStat[]), (args[1] as integer), true, false, false));
+    pokemon.scene.unshiftPhase(new StatStageChangePhase(pokemon.scene, pokemon.getBattlerIndex(), true, (args[0] as BattleStat[]), (args[1] as integer), true, false, false));
```
nit
```suggestion
    pokemon.scene.unshiftPhase(new StatStageChangePhase(pokemon.scene, pokemon.getBattlerIndex(), true, (args[0] as BattleStat[]), (args[1] as number), true, false, false));
```

torranx reviewed (commented):
initial review only 😄

xsn34kzx commented:
Just to update, `Guard Swap` and `Power Swap` are now on the docket to be added, and the messages that follow all the moves being implemented will also get added (I forgot they had unique effect messages).

frutescens commented:
Do you have an idea how moves like Burning Jealousy/Alluring Voice would be handled in this PR? Just wondering since there's an active PR #3508 implementing them

xsn34kzx commented:
> Do you have an idea how moves like Burning Jealousy/Alluring Voice would be handled in this PR? Just wondering since there's an active PR #3508 implementing them

@frutescens I think besides the semantic differences introduced in this PR (`battleStats` -> `statStages`), that PR looks like it *should* be completely safe from anything done here.

xsn34kzx commented:
The move effect messages for each more were sourced from Scarlet/Violet via PokeCorpus:
- [Power Swap](https://abcboy101.github.io/poke-corpus/#id=sv.btl_set.BTL_STRID_SET_PowerSwap)
- [Guard Swap](https://abcboy101.github.io/poke-corpus/#id=sv.btl_set.BTL_STRID_SET_GuardSwap)
- [Speed Swap](https://abcboy101.github.io/poke-corpus/#id=sv.btl_set.BTL_STRID_SET_SpeedSwap)
- [Power Split](https://abcboy101.github.io/poke-corpus/#id=sv.btl_set.BTL_STRID_SET_PowerShare)
- [Guard Split](https://abcboy101.github.io/poke-corpus/#id=sv.btl_set.BTL_STRID_SET_GuardShare)

xsn34kzx commented:
Localization entries were added; while adding them, I tried my best to use PokeCorpus for comparison on how things like stat names should fit as they do in the English example, but it's definitely not perfect.

flx-sta commented on src/data/arena-tag.ts:
```diff
@@ -17,7 +17,7 @@ import { Moves } from "#enums/moves";
 import { PokemonHealPhase } from "#app/phases/pokemon-heal-phase.js";
 import { ShowAbilityPhase } from "#app/phases/show-ability-phase.js";
-import { StatChangePhase } from "#app/phases/stat-change-phase.js";
+import { StatStageChangePhase } from "#app/phases/stat-stage-change-phase.js";
```
```suggestion
import { StatStageChangePhase } from "#app/phases/stat-stage-change-phase";
```

flx-sta commented on src/data/battler-tags.ts:
```diff
@@ -1,28 +1,27 @@
 import { PokemonHealPhase } from "#app/phases/pokemon-heal-phase.js";
 import { ShowAbilityPhase } from "#app/phases/show-ability-phase.js";
-import { StatChangePhase, StatChangeCallback } from "#app/phases/stat-change-phase.js";
+import { StatStageChangePhase, StatStageChangeCallback } from "#app/phases/stat-stage-change-phase.js";
```
```suggestion
import { StatStageChangePhase, StatStageChangeCallback } from "#app/phases/stat-stage-change-phase";
```

flx-sta commented on src/data/move.ts:
```diff
@@ -27,13 +25,14 @@ import { Biome } from "#enums/biome";
 import { NewBattlePhase } from "#app/phases/new-battle-phase.js";
 import { PokemonHealPhase } from "#app/phases/pokemon-heal-phase.js";
-import { StatChangePhase } from "#app/phases/stat-change-phase.js";
+import { StatStageChangePhase } from "#app/phases/stat-stage-change-phase.js";
```
```suggestion
import { StatStageChangePhase } from "#app/phases/stat-stage-change-phase";
```

flx-sta commented on src/field/pokemon.ts:
```diff
@@ -56,7 +54,7 @@ import { LearnMovePhase } from "#app/phases/learn-move-phase.js";
 import { MoveEndPhase } from "#app/phases/move-end-phase.js";
 import { ObtainStatusEffectPhase } from "#app/phases/obtain-status-effect-phase.js";
-import { StatChangePhase } from "#app/phases/stat-change-phase.js";
+import { StatStageChangePhase } from "#app/phases/stat-stage-change-phase.js";
```
```suggestion
import { StatStageChangePhase } from "#app/phases/stat-stage-change-phase";
```

flx-sta commented on src/locales/fr/modifier.json:
```diff
@@ -3,7 +3,7 @@
   "hitHealApply": "Les PV de {{pokemonNameWithAffix}}\nsont un peu restaurés par le {{typeName}} !",
   "pokemonInstantReviveApply": "{{pokemonNameWithAffix}} a repris connaissance\navec sa {{typeName}} et est prêt à se battre de nouveau !",
-  "pokemonResetNegativeStatStageApply": "{{pokemonNameWithAffix}}'s lowered stats were restored\nby its {{typeName}}!",
+  "resetNegativeStatStageApply": "{{pokemonNameWithAffix}}'s lowered stats were restored\nby its {{typeName}}!",
```
```suggestion
```
No more adding fake translation. This isn't french (as far as I'm aware)

flx-sta commented on src/locales/pt_BR/move-trigger.json:
```diff
@@ -3,6 +3,10 @@
   "cutHpPowerUpMove": "{{pokemonName}} diminuiu seus PS para aumentar o poder do ataque!",
   "absorbedElectricity": "{{pokemonName}} absorveu eletricidade!",
   "switchedStatChanges": "{{pokemonName}} trocou as mudanças de atributo com o alvo!",
+  "switchedTwoStatChanges": "{{pokemonName}} switched all changes to its {{firstStat}}\nand {{secondStat}} with its target!",
+  "switchedStat": "{{pokemonName}} switched {{stat}} with its target!",
+  "sharedGuard": "{{pokemonName}} shared its guard with the target!",
+  "sharedPower": "{{pokemonName}} shared its power with the target!",
   "goingAllOutForAttack": "{{pokemonName}} está arriscando tudo nesse ataque!",
```
```suggestion
  "switchedStatChanges": "{{pokemonName}} trocou as mudanças de atributo com o alvo!",
  "goingAllOutForAttack": "{{pokemonName}} está arriscando tudo nesse ataque!",
```

No more adding "fake" translations

flx-sta commented on src/phases/faint-phase.ts:
```diff
@@ -3,7 +3,7 @@ import { BattlerIndex, BattleType } from "#app/battle.js";
 import { BattlerTagLapseType } from "#app/data/battler-tags.js";
 import { battleSpecDialogue } from "#app/data/dialogue.js";
-import { allMoves, PostVictoryStatChangeAttr } from "#app/data/move.js";
+import { allMoves, PostVictoryStatStageChangeAttr } from "#app/data/move.js";
```
```suggestion
import { allMoves, PostVictoryStatStageChangeAttr } from "#app/data/move";
```

flx-sta commented on src/phases/stat-stage-change-phase.ts:
```diff
@@ -0,0 +1,222 @@
+import BattleScene from "#app/battle-scene.js";
+import { BattlerIndex } from "#app/battle.js";
+import { applyPreStatStageChangeAbAttrs, ProtectStatAbAttr, applyAbAttrs, StatStageChangeMultiplierAbAttr, StatStageChangeCopyAbAttr, applyPostStatStageChangeAbAttrs, PostStatStageChangeAbAttr } from "#app/data/ability.js";
+import { MistTag, ArenaTagSide } from "#app/data/arena-tag.js";
+import Pokemon from "#app/field/pokemon.js";
+import { getPokemonNameWithAffix } from "#app/messages.js";
+import { ResetNegativeStatStageModifier } from "#app/modifier/modifier.js";
+import { handleTutorial, Tutorial } from "#app/tutorial.js";
+import i18next from "i18next";
+import * as Utils from "#app/utils.js";
+import { PokemonPhase } from "./pokemon-phase";
+import { Stat, type BattleStat, getStatKey, getStatStageChangeDescriptionKey } from "#enums/stat";
```
```suggestion
import BattleScene from "#app/battle-scene.js";
import { BattlerIndex } from "#app/battle.js";
import { applyPreStatStageChangeAbAttrs, ProtectStatAbAttr, applyAbAttrs, StatStageChangeMultiplierAbAttr, StatStageChangeCopyAbAttr, applyPostStatStageChangeAbAttrs, PostStatStageChangeAbAttr } from "#app/data/ability.js";
import { MistTag, ArenaTagSide } from "#app/data/arena-tag.js";
import Pokemon from "#app/field/pokemon.js";
import { getPokemonNameWithAffix } from "#app/messages.js";
import { ResetNegativeStatStageModifier } from "#app/modifier/modifier.js";
import { handleTutorial, Tutorial } from "#app/tutorial";
import i18next from "i18next";
import * as Utils from "#app/utils";
import { PokemonPhase } from "./pokemon-phase";
import { Stat, type BattleStat, getStatKey, getStatStageChangeDescriptionKey } from "#enums/stat";
```

flx-sta commented on src/phases/stat-stage-change-phase.ts:
```diff
@@ -0,0 +1,222 @@
+      const existingPhase = this.scene.findPhase(p => p instanceof StatStageChangePhase && p.battlerIndex === this.battlerIndex);
+      if (!(existingPhase instanceof StatStageChangePhase)) {
+        // Apply White Herb if needed
+        const whiteHerb = this.scene.applyModifier(ResetNegativeStatStageModifier, this.player, pokemon) as ResetNegativeStatStageModifier;
```
wasn't whiteHerb removed?

xsn34kzx replied:
It was removed from the item pool, but it's functionality code was left in for usage in the future, so I took the opportunity to clean up the code with the new stuff introduced here.

flx-sta commented on src/test/abilities/beast_boost.test.ts:
```diff
@@ -0,0 +1,97 @@
+import Phaser from "phaser";
+import { afterEach, beforeAll, beforeEach, describe, expect, it, vi } from "vitest";
+import { SPLASH_ONLY } from "../utils/testUtils";
+import { EnemyCommandPhase } from "#app/phases/enemy-command-phase.js";
```
```suggestion
import { EnemyCommandPhase } from "#app/phases/enemy-command-phase";
```

flx-sta commented on src/test/abilities/disguise.test.ts:
```diff
@@ -9,6 +7,8 @@ import { Mode } from "#app/ui/ui";
 import { toDmgValue } from "#app/utils";
 import { Moves } from "#enums/moves";
 import { Species } from "#enums/species";
+import { StatusEffect } from "#app/data/status-effect.js";
```
```suggestion
import { StatusEffect } from "#app/data/status-effect";
```

flx-sta commented on src/test/abilities/moxie.test.ts:
```diff
@@ -1,14 +1,15 @@
-
+import { SPLASH_ONLY } from "../utils/testUtils";
+import { BattlerIndex } from "#app/battle";
+import { EnemyCommandPhase } from "#app/phases/enemy-command-phase.js";
```
```suggestion
import { EnemyCommandPhase } from "#app/phases/enemy-command-phase";
```

flx-sta commented on src/test/abilities/moxie.test.ts:
```diff
@@ -1,14 +1,15 @@
-
+import { SPLASH_ONLY } from "../utils/testUtils";
+import { BattlerIndex } from "#app/battle";
+import { EnemyCommandPhase } from "#app/phases/enemy-command-phase.js";
+import { VictoryPhase } from "#app/phases/victory-phase.js";
```
```suggestion
import { EnemyCommandPhase } from "#app/phases/enemy-command-phase";
import { VictoryPhase } from "#app/phases/victory-phase";
```

flx-sta commented on src/test/abilities/mycelium_might.test.ts:
```diff
@@ -1,10 +1,10 @@
-import { BattleStat } from "#app/data/battle-stat";
-import { MovePhase } from "#app/phases/move-phase";
-import { TurnEndPhase } from "#app/phases/turn-end-phase";
+import { MovePhase } from "#app/phases/move-phase.js";
+import { TurnEndPhase } from "#app/phases/turn-end-phase.js";
```
```suggestion
import { MovePhase } from "#app/phases/move-phase";
import { TurnEndPhase } from "#app/phases/turn-end-phase";
```

flx-sta commented on src/test/abilities/sheer_force.test.ts:
```diff
@@ -1,7 +1,7 @@
-import { Stat } from "#app/data/pokemon-stat";
-import { MoveEffectPhase } from "#app/phases/move-effect-phase";
+import { Stat } from "#enums/stat";
+import { MoveEffectPhase } from "#app/phases/move-effect-phase.js";
```
```suggestion
import { MoveEffectPhase } from "#app/phases/move-effect-phase";
```

flx-sta commented on src/test/abilities/volt_absorb.test.ts:
```diff
@@ -1,5 +1,5 @@
-import { BattleStat } from "#app/data/battle-stat";
-import { TurnEndPhase } from "#app/phases/turn-end-phase";
+import { Stat } from "#enums/stat";
+import { TurnEndPhase } from "#app/phases/turn-end-phase.js";
```
```suggestion
import { TurnEndPhase } from "#app/phases/turn-end-phase";
```

flx-sta commented on src/test/abilities/wind_rider.test.ts:
```diff
@@ -1,9 +1,9 @@
-import { BattleStat } from "#app/data/battle-stat";
-import { TurnEndPhase } from "#app/phases/turn-end-phase";
+import { Stat } from "#enums/stat";
+import { TurnEndPhase } from "#app/phases/turn-end-phase.js";
```
```suggestion
import { TurnEndPhase } from "#app/phases/turn-end-phase";
```

flx-sta commented on src/test/battlerTags/octolock.test.ts:
```diff
@@ -1,16 +1,16 @@
-import { StatChangePhase } from "#app/phases/stat-change-phase";
 import { describe, expect, it, vi } from "vitest";
+import Pokemon from "#app/field/pokemon.js";
+import { BattlerTag, BattlerTagLapseType, OctolockTag, TrappedTag } from "#app/data/battler-tags.js";
+import { StatStageChangePhase } from "#app/phases/stat-stage-change-phase";
+import { BattlerTagType } from "#app/enums/battler-tag-type.js";
```
```suggestion
import { BattlerTag, BattlerTagLapseType, OctolockTag, TrappedTag } from "#app/data/battler-tags";
import { StatStageChangePhase } from "#app/phases/stat-stage-change-phase";
import { BattlerTagType } from "#app/enums/battler-tag-type";
```

flx-sta commented on src/test/battlerTags/stockpiling.test.ts:
```diff
@@ -1,18 +1,18 @@
-import * as messages from "#app/messages";
-import { StatChangePhase } from "#app/phases/stat-change-phase";
 import { beforeEach, describe, expect, it, vi } from "vitest";
+import Pokemon, { PokemonSummonData } from "#app/field/pokemon.js";
+import { StockpilingTag } from "#app/data/battler-tags.js";
+import { Stat } from "#enums/stat";
+import * as messages from "#app/messages.js";
+import { StatStageChangePhase } from "#app/phases/stat-stage-change-phase";
```
```suggestion
import Pokemon, { PokemonSummonData } from "#app/field/pokemon";
import { StockpilingTag } from "#app/data/battler-tags";
import { Stat } from "#enums/stat";
import * as messages from "#app/messages";
import { StatStageChangePhase } from "#app/phases/stat-stage-change-phase";
```

flx-sta commented on src/test/items/dire_hit.test.ts:
```diff
@@ -0,0 +1,97 @@
+import { TurnEndPhase } from "#app/phases/turn-end-phase.js";
+import { Moves } from "#enums/moves";
+import { Species } from "#enums/species";
+import GameManager from "#test/utils/gameManager";
+import Phase from "phaser";
+import { afterEach, beforeAll, beforeEach, describe, expect, it, vi } from "vitest";
+import { SPLASH_ONLY } from "../utils/testUtils";
+import { BattleEndPhase } from "#app/phases/battle-end-phase.js";
+import { TempCritBoosterModifier } from "#app/modifier/modifier.js";
+import { Mode } from "#app/ui/ui.js";
+import ModifierSelectUiHandler from "#app/ui/modifier-select-ui-handler.js";
+import { Button } from "#app/enums/buttons.js";
+import { CommandPhase } from "#app/phases/command-phase.js";
+import { NewBattlePhase } from "#app/phases/new-battle-phase.js";
+import { TurnInitPhase } from "#app/phases/turn-init-phase.js";
```
```suggestion
import { TurnEndPhase } from "#app/phases/turn-end-phase";
import { Moves } from "#enums/moves";
import { Species } from "#enums/species";
import GameManager from "#test/utils/gameManager";
import Phase from "phaser";
import { afterEach, beforeAll, beforeEach, describe, expect, it, vi } from "vitest";
import { SPLASH_ONLY } from "../utils/testUtils";
import { BattleEndPhase } from "#app/phases/battle-end-phase";
import { TempCritBoosterModifier } from "#app/modifier/modifier";
import { Mode } from "#app/ui/ui.js";
import ModifierSelectUiHandler from "#app/ui/modifier-select-ui-handler";
import { Button } from "#app/enums/buttons.js";
import { CommandPhase } from "#app/phases/command-phase";
import { NewBattlePhase } from "#app/phases/new-battle-phase";
import { TurnInitPhase } from "#app/phases/turn-init-phase";
```

flx-sta commented on src/test/items/leek.test.ts:
```diff
@@ -1,7 +1,4 @@
-import { CritBoosterModifier } from "#app/modifier/modifier";
-import { modifierTypes } from "#app/modifier/modifier-type";
-import { MoveEffectPhase } from "#app/phases/move-effect-phase";
+import { TurnEndPhase } from "#app/phases/turn-end-phase.js";
```
```suggestion
import { TurnEndPhase } from "#app/phases/turn-end-phase";
```

flx-sta commented on src/test/items/scope_lens.test.ts:
```diff
@@ -1,13 +1,10 @@
-import { modifierTypes } from "#app/modifier/modifier-type";
-import { MoveEffectPhase } from "#app/phases/move-effect-phase";
-import * as Utils from "#app/utils";
+import { TurnEndPhase } from "#app/phases/turn-end-phase.js";
```
```suggestion
import { TurnEndPhase } from "#app/phases/turn-end-phase";
```

flx-sta commented on src/test/items/temp_stat_stage_booster.test.ts:
```diff
@@ -0,0 +1,174 @@
+import { BATTLE_STATS, Stat } from "#enums/stat";
+import GameManager from "#test/utils/gameManager";
+import { Species } from "#enums/species";
+import Phase from "phaser";
+import { afterEach, beforeAll, beforeEach, describe, expect, it, vi } from "vitest";
+import { Moves } from "#app/enums/moves.js";
+import { TurnEndPhase } from "#app/phases/turn-end-phase.js";
+import { SPLASH_ONLY } from "../utils/testUtils";
+import { Abilities } from "#app/enums/abilities.js";
+import { TempStatStageBoosterModifier } from "#app/modifier/modifier.js";
+import { Mode } from "#app/ui/ui.js";
+import { Button } from "#app/enums/buttons.js";
+import { CommandPhase } from "#app/phases/command-phase.js";
+import { NewBattlePhase } from "#app/phases/new-battle-phase.js";
+import ModifierSelectUiHandler from "#app/ui/modifier-select-ui-handler.js";
+import { TurnInitPhase } from "#app/phases/turn-init-phase.js";
+import { BattleEndPhase } from "#app/phases/battle-end-phase.js";
+import { EnemyCommandPhase } from "#app/phases/enemy-command-phase.js";
+
```
```suggestion
import { BATTLE_STATS, Stat } from "#enums/stat";
import GameManager from "#test/utils/gameManager";
import { Species } from "#enums/species";
import Phase from "phaser";
import { afterEach, beforeAll, beforeEach, describe, expect, it, vi } from "vitest";
import { Moves } from "#app/enums/moves";
import { TurnEndPhase } from "#app/phases/turn-end-phase";
import { SPLASH_ONLY } from "../utils/testUtils";
import { Abilities } from "#app/enums/abilities";
import { TempStatStageBoosterModifier } from "#app/modifier/modifier";
import { Mode } from "#app/ui/ui";
import { Button } from "#app/enums/buttons";
import { CommandPhase } from "#app/phases/command-phase";
import { NewBattlePhase } from "#app/phases/new-battle-phase";
import ModifierSelectUiHandler from "#app/ui/modifier-select-ui-handler";
import { TurnInitPhase } from "#app/phases/turn-init-phase";
import { BattleEndPhase } from "#app/phases/battle-end-phase";
import { EnemyCommandPhase } from "#app/phases/enemy-command-phase";

```

flx-sta commented on src/test/moves/clangorous_soul.test.ts:
```diff
@@ -1,20 +1,19 @@
+import { afterEach, beforeAll, beforeEach, describe, expect, it } from "vitest";
+import Phaser from "phaser";
+import GameManager from "#test/utils/gameManager";
+import { TurnEndPhase } from "#app/phases/turn-end-phase.js";
```
```suggestion
import { TurnEndPhase } from "#app/phases/turn-end-phase";
```

flx-sta commented on src/test/moves/crafty_shield.test.ts:
```diff
@@ -1,13 +1,13 @@
+import { Abilities } from "#enums/abilities";
+import { Moves } from "#enums/moves";
+import { Stat } from "#enums/stat";
+import { BattlerTagType } from "#app/enums/battler-tag-type.js";
+import { BerryPhase } from "#app/phases/berry-phase.js";
+import { CommandPhase } from "#app/phases/command-phase.js";
```
```suggestion
import { BattlerTagType } from "#app/enums/battler-tag-type";
import { BerryPhase } from "#app/phases/berry-phase";
import { CommandPhase } from "#app/phases/command-phase";
```

flx-sta commented on src/test/moves/fissure.test.ts:
```diff
@@ -1,5 +1,5 @@
-import { BattleStat } from "#app/data/battle-stat";
-import { Species } from "#app/enums/species";
+import { Stat } from "#enums/stat";
+import { Species } from "#app/enums/species.js";
```
```suggestion
import { Species } from "#app/enums/species";
```

flx-sta commented on src/test/moves/freezy_frost.test.ts:
```diff
@@ -1,82 +1,61 @@
 import { afterEach, beforeAll, beforeEach, describe, expect, it, vi } from "vitest";
+import { SPLASH_ONLY } from "#test/utils/testUtils";
+import { allMoves } from "#app/data/move.js";
+import { TurnInitPhase } from "#app/phases/turn-init-phase.js";
```
```suggestion
import { TurnInitPhase } from "#app/phases/turn-init-phase";
```

flx-sta commented on src/test/moves/growth.test.ts:
```diff
@@ -1,14 +1,13 @@
 import { afterEach, beforeAll, beforeEach, describe, expect, it } from "vitest";
-
+import { SPLASH_ONLY } from "../utils/testUtils";
+import { EnemyCommandPhase } from "#app/phases/enemy-command-phase.js";
+import { TurnInitPhase } from "#app/phases/turn-init-phase.js";
```
```suggestion
import { EnemyCommandPhase } from "#app/phases/enemy-command-phase";
import { TurnInitPhase } from "#app/phases/turn-init-phase";
```

flx-sta commented on src/test/moves/guard_swap.test.ts:
```diff
@@ -0,0 +1,63 @@
+import { Moves } from "#enums/moves";
+import { Stat } from "#enums/stat";
+import { Abilities } from "#enums/abilities";
+import { MoveEndPhase } from "#app/phases/move-end-phase.js";
```
```suggestion
import { MoveEndPhase } from "#app/phases/move-end-phase";
```

flx-sta commented on src/test/moves/haze.test.ts:
```diff
@@ -1,13 +1,12 @@
 import Phaser from "phaser";
 import { afterEach, beforeAll, beforeEach, describe, expect, it } from "vitest";
+import { SPLASH_ONLY } from "#test/utils/testUtils";
+import { TurnInitPhase } from "#app/phases/turn-init-phase.js";
```
```suggestion
import { TurnInitPhase } from "#app/phases/turn-init-phase";
```

flx-sta commented on src/test/moves/make_it_rain.test.ts:
```diff
@@ -1,13 +1,13 @@
 import Phaser from "phaser";
 import { afterEach, beforeAll, beforeEach, describe, expect, it } from "vitest";
+import { SPLASH_ONLY } from "#test/utils/testUtils";
+import { MoveEndPhase } from "#app/phases/move-end-phase.js";
```
```suggestion
import { MoveEndPhase } from "#app/phases/move-end-phase";
```

flx-sta commented on src/test/moves/mat_block.test.ts:
```diff
@@ -1,13 +1,13 @@
+import { Abilities } from "#enums/abilities";
+import { Moves } from "#enums/moves";
+import { Stat } from "#enums/stat";
+import { BerryPhase } from "#app/phases/berry-phase.js";
+import { CommandPhase } from "#app/phases/command-phase.js";
+import { TurnEndPhase } from "#app/phases/turn-end-phase.js";
```
```suggestion
import { BerryPhase } from "#app/phases/berry-phase";
import { CommandPhase } from "#app/phases/command-phase";
import { TurnEndPhase } from "#app/phases/turn-end-phase";
```

flx-sta commented on src/test/moves/parting_shot.test.ts:
```diff
@@ -1,14 +1,14 @@
 import Phaser from "phaser";
 import { afterEach, beforeAll, beforeEach, describe, expect, it, test } from "vitest";
 import GameManager from "../utils/gameManager";
+import { Stat } from "#enums/stat";
+import { BerryPhase } from "#app/phases/berry-phase.js";
+import { FaintPhase } from "#app/phases/faint-phase.js";
+import { MessagePhase } from "#app/phases/message-phase.js";
+import { TurnInitPhase } from "#app/phases/turn-init-phase.js";
 import { SPLASH_ONLY } from "../utils/testUtils";
```
```suggestion
import { Stat } from "#enums/stat";
import { BerryPhase } from "#app/phases/berry-phase";
import { FaintPhase } from "#app/phases/faint-phase";
import { MessagePhase } from "#app/phases/message-phase";
import { TurnInitPhase } from "#app/phases/turn-init-phase";
import { SPLASH_ONLY } from "../utils/testUtils";
```

flx-sta commented on src/test/moves/power_swap.test.ts:
```diff
@@ -0,0 +1,62 @@
+import { Moves } from "#enums/moves";
+import { Stat } from "#enums/stat";
+import { Abilities } from "#enums/abilities";
+import { MoveEndPhase } from "#app/phases/move-end-phase.js";
```
```suggestion
import { MoveEndPhase } from "#app/phases/move-end-phase";
```

flx-sta commented on src/test/moves/protect.test.ts:
```diff
@@ -1,13 +1,13 @@
+import { Moves } from "#enums/moves";
+import { Stat } from "#enums/stat";
+import { allMoves } from "#app/data/move.js";
+import { ArenaTagSide, ArenaTrapTag } from "#app/data/arena-tag.js";
+import { BerryPhase } from "#app/phases/berry-phase.js";
```
```suggestion
import { ArenaTagSide, ArenaTrapTag } from "#app/data/arena-tag";
import { BerryPhase } from "#app/phases/berry-phase";
```

flx-sta commented on src/test/moves/quick_guard.test.ts:
```diff
@@ -1,12 +1,12 @@
+import { Abilities } from "#enums/abilities";
+import { Moves } from "#enums/moves";
+import { Stat } from "#enums/stat";
+import { BerryPhase } from "#app/phases/berry-phase.js";
+import { CommandPhase } from "#app/phases/command-phase.js";
```
```suggestion
import { BerryPhase } from "#app/phases/berry-phase";
import { CommandPhase } from "#app/phases/command-phase";
```

flx-sta commented on src/test/moves/spit_up.test.ts:
```diff
@@ -1,22 +1,24 @@
-import { MoveResult, TurnMove } from "#app/field/pokemon";
-import { MovePhase } from "#app/phases/move-phase";
-import { TurnInitPhase } from "#app/phases/turn-init-phase";
+import { Stat } from "#enums/stat";
+import { StockpilingTag } from "#app/data/battler-tags.js";
+import { allMoves } from "#app/data/move.js";
+import { BattlerTagType } from "#app/enums/battler-tag-type.js";
+import { MoveResult, TurnMove } from "#app/field/pokemon.js";
+import GameManager from "#test/utils/gameManager";
```
```suggestion
import { Stat } from "#enums/stat";
import { StockpilingTag } from "#app/data/battler-tags";
import { allMoves } from "#app/data/move";
import { BattlerTagType } from "#app/enums/battler-tag-type";
import { MoveResult, TurnMove } from "#app/field/pokemon";
import GameManager from "#test/utils/gameManager";
```

flx-sta commented on src/test/moves/spit_up.test.ts:
```diff
@@ -1,22 +1,24 @@
-import { SPLASH_ONLY } from "#test/utils/testUtils";
 import Phaser from "phaser";
 import { afterEach, beforeAll, beforeEach, describe, expect, it, vi } from "vitest";
+import { SPLASH_ONLY } from "#test/utils/testUtils";
+import { MovePhase } from "#app/phases/move-phase.js";
+import { TurnInitPhase } from "#app/phases/turn-init-phase.js";
```
```suggestion
import { SPLASH_ONLY } from "#test/utils/testUtils";
import { MovePhase } from "#app/phases/move-phase";
import { TurnInitPhase } from "#app/phases/turn-init-phase";
```

flx-sta commented on src/test/moves/spotlight.test.ts:
```diff
@@ -1,5 +1,5 @@
-import { BattlerIndex } from "#app/battle";
-import { Stat } from "#app/data/pokemon-stat";
+import { BattlerIndex } from "#app/battle.js";
```
```suggestion
import { BattlerIndex } from "#app/battle";
```

flx-sta commented on src/test/moves/tail_whip.test.ts:
```diff
@@ -1,12 +1,13 @@
 import Phaser from "phaser";
 import { afterEach, beforeAll, beforeEach, describe, expect, it } from "vitest";
+import { SPLASH_ONLY } from "../utils/testUtils";
+import { EnemyCommandPhase } from "#app/phases/enemy-command-phase.js";
+import { TurnInitPhase } from "#app/phases/turn-init-phase.js";
```
```suggestion
import { EnemyCommandPhase } from "#app/phases/enemy-command-phase";
import { TurnInitPhase } from "#app/phases/turn-init-phase";
```

flx-sta commented on src/test/moves/tidy_up.test.ts:
```diff
@@ -1,5 +1,5 @@
-import { BattleStat } from "#app/data/battle-stat";
-import { ArenaTagType } from "#app/enums/arena-tag-type";
+import { Stat } from "#enums/stat";
+import { ArenaTagType } from "#app/enums/arena-tag-type.js";
```
```suggestion
import { ArenaTagType } from "#app/enums/arena-tag-type";
```

flx-sta commented on src/test/moves/wide_guard.test.ts:
```diff
@@ -1,12 +1,12 @@
+import { Abilities } from "#enums/abilities";
+import { Moves } from "#enums/moves";
+import { Stat } from "#enums/stat";
+import { BerryPhase } from "#app/phases/berry-phase.js";
+import { CommandPhase } from "#app/phases/command-phase.js";
```
```suggestion
import { BerryPhase } from "#app/phases/berry-phase";
import { CommandPhase } from "#app/phases/command-phase";
```

flx-sta commented on src/test/utils/phaseInterceptor.ts:
```diff
@@ -1,41 +1,41 @@
-import { UnavailablePhase } from "#app/phases/unavailable-phase";
-import { VictoryPhase } from "#app/phases/victory-phase";
 import ErrorInterceptor from "#app/test/utils/errorInterceptor";
+import { BattleEndPhase } from "#app/phases/battle-end-phase.js";
+import { BerryPhase } from "#app/phases/berry-phase.js";
+import { CheckSwitchPhase } from "#app/phases/check-switch-phase.js";
+import { CommandPhase } from "#app/phases/command-phase.js";
+import { DamagePhase } from "#app/phases/damage-phase.js";
+import { EggLapsePhase } from "#app/phases/egg-lapse-phase.js";
+import { EncounterPhase } from "#app/phases/encounter-phase.js";
+import { EnemyCommandPhase } from "#app/phases/enemy-command-phase.js";
+import { FaintPhase } from "#app/phases/faint-phase.js";
+import { LoginPhase } from "#app/phases/login-phase.js";
+import { MessagePhase } from "#app/phases/message-phase.js";
+import { MoveEffectPhase } from "#app/phases/move-effect-phase.js";
+import { MoveEndPhase } from "#app/phases/move-end-phase.js";
+import { MovePhase } from "#app/phases/move-phase.js";
+import { NewBattlePhase } from "#app/phases/new-battle-phase.js";
+import { NextEncounterPhase } from "#app/phases/next-encounter-phase.js";
+import { PostSummonPhase } from "#app/phases/post-summon-phase.js";
+import { QuietFormChangePhase } from "#app/phases/quiet-form-change-phase.js";
+import { SelectGenderPhase } from "#app/phases/select-gender-phase.js";
+import { SelectModifierPhase } from "#app/phases/select-modifier-phase.js";
+import { SelectStarterPhase } from "#app/phases/select-starter-phase.js";
+import { SelectTargetPhase } from "#app/phases/select-target-phase.js";
+import { ShinySparklePhase } from "#app/phases/shiny-sparkle-phase.js";
+import { ShowAbilityPhase } from "#app/phases/show-ability-phase.js";
+import { StatStageChangePhase } from "#app/phases/stat-stage-change-phase";
+import { SummonPhase } from "#app/phases/summon-phase.js";
+import { SwitchPhase } from "#app/phases/switch-phase.js";
+import { SwitchSummonPhase } from "#app/phases/switch-summon-phase.js";
+import { TitlePhase } from "#app/phases/title-phase.js";
+import { ToggleDoublePositionPhase } from "#app/phases/toggle-double-position-phase.js";
+import { TurnEndPhase } from "#app/phases/turn-end-phase.js";
+import { TurnInitPhase } from "#app/phases/turn-init-phase.js";
+import { TurnStartPhase } from "#app/phases/turn-start-phase.js";
+import { UnavailablePhase } from "#app/phases/unavailable-phase.js";
+import { VictoryPhase } from "#app/phases/victory-phase.js";
+import { PartyHealPhase } from "#app/phases/party-heal-phase.js";
```
```suggestion
```suggestion
import { BattleEndPhase } from "#app/phases/battle-end-phase";
import { BerryPhase } from "#app/phases/berry-phase";
import { CheckSwitchPhase } from "#app/phases/check-switch-phase";
import { CommandPhase } from "#app/phases/command-phase";
import { DamagePhase } from "#app/phases/damage-phase";
import { EggLapsePhase } from "#app/phases/egg-lapse-phase";
import { EncounterPhase } from "#app/phases/encounter-phase";
import { EnemyCommandPhase } from "#app/phases/enemy-command-phase";
import { FaintPhase } from "#app/phases/faint-phase";
import { LoginPhase } from "#app/phases/login-phase";
import { MessagePhase } from "#app/phases/message-phase";
import { MoveEffectPhase } from "#app/phases/move-effect-phase";
import { MoveEndPhase } from "#app/phases/move-end-phase";
import { MovePhase } from "#app/phases/move-phase";
import { NewBattlePhase } from "#app/phases/new-battle-phase";
import { NextEncounterPhase } from "#app/phases/next-encounter-phase";
import { PostSummonPhase } from "#app/phases/post-summon-phase";
import { QuietFormChangePhase } from "#app/phases/quiet-form-change-phase";
import { SelectGenderPhase } from "#app/phases/select-gender-phase";
import { SelectModifierPhase } from "#app/phases/select-modifier-phase";
import { SelectStarterPhase } from "#app/phases/select-starter-phase";
import { SelectTargetPhase } from "#app/phases/select-target-phase";
import { ShinySparklePhase } from "#app/phases/shiny-sparkle-phase";
import { ShowAbilityPhase } from "#app/phases/show-ability-phase";
import { StatStageChangePhase } from "#app/phases/stat-stage-change-phase";
import { SummonPhase } from "#app/phases/summon-phase";
import { SwitchPhase } from "#app/phases/switch-phase";
import { SwitchSummonPhase } from "#app/phases/switch-summon-phase";
import { TitlePhase } from "#app/phases/title-phase";
import { ToggleDoublePositionPhase } from "#app/phases/toggle-double-position-phase";
import { TurnEndPhase } from "#app/phases/turn-end-phase";
import { TurnInitPhase } from "#app/phases/turn-init-phase";
import { TurnStartPhase } from "#app/phases/turn-start-phase";
import { UnavailablePhase } from "#app/phases/unavailable-phase";
import { VictoryPhase } from "#app/phases/victory-phase";
import { PartyHealPhase } from "#app/phases/party-heal-phase";
```
```

flx-sta commented on src/ui/stats-container.ts:
```diff
@@ -1,7 +1,8 @@
 import BattleScene from "../battle-scene";
-import { Stat, getStatName } from "../data/pokemon-stat";
 import { TextStyle, addBBCodeTextObject, addTextObject, getTextColor } from "./text";
+import { PERMANENT_STATS, getStatKey } from "#app/enums/stat.js";
```
```suggestion
import { PERMANENT_STATS, getStatKey } from "#app/enums/stat";
```

flx-sta commented on src/ui/summary-ui-handler.ts:
```diff
@@ -23,6 +22,7 @@ import { Ability } from "../data/ability.js";
 import i18next from "i18next";
 import {modifierSortFunc} from "../modifier/modifier";
 import { PlayerGender } from "#enums/player-gender";
+import { Stat, PERMANENT_STATS, getStatKey } from "#app/enums/stat.js";
```
```suggestion
import { Stat, PERMANENT_STATS, getStatKey } from "#app/enums/stat";
```

flx-sta commented:
Did a pure import check on `.js` endings. 
Idk how to really review this except play testing I guess. Don't have the time right now though

xsn34kzx commented on src/modifier/modifier.ts:
```diff
@@ -3,14 +3,13 @@ import BattleScene from "../battle-scene";
 import { MAX_PER_TYPE_POKEBALLS, PokeballType } from "../data/pokeball";
 import Pokemon, { PlayerPokemon } from "../field/pokemon";
-import { Stat } from "../data/pokemon-stat";
+import { Stat } from "#enums/stat";
```
```suggestion
```
NIT: Remove enum import here to another line

xsn34kzx commented on src/modifier/modifier.ts:
```diff
@@ -23,6 +22,7 @@ import Overrides from "#app/overrides";
 import { ModifierType, modifierTypes } from "./modifier-type";
 import { Command } from "#app/ui/command-ui-handler";
 import { Species } from "#enums/species";
+import { type PermanentStat, type TempBattleStat, BATTLE_STATS, TEMP_BATTLE_STATS  } from "#app/enums/stat";
```
```suggestion
import { Stat, type PermanentStat, type TempBattleStat, BATTLE_STATS, TEMP_BATTLE_STATS  } from "#app/enums/stat";
```
Add back removed `Stat` from previous comment.

xsn34kzx commented on src/test/abilities/intimidate.test.ts:
```diff
@@ -1,21 +1,15 @@
 import { SPLASH_ONLY } from "#test/utils/testUtils";
-import Phaser from "phaser";
-import { afterEach, beforeAll, beforeEach, describe, expect, it } from "vitest";
+import { CommandPhase } from "#app/phases/command-phase.js";
+import { TurnInitPhase } from "#app/phases/turn-init-phase.js";
```
```suggestion
import { CommandPhase } from "#app/phases/command-phase";
import { TurnInitPhase } from "#app/phases/turn-init-phase";
```
NIT: `.js` import

xsn34kzx reviewed (commented):
Did another once-over while batching suggestions.

innerthunder reviewed (dismissed):
Meant to do this earlier... I did a full Classic run on this, and all the issues I observed have since been fixed. Let's break some PRs 😈

frutescens reviewed (commented):
It would be good to have tests checking that Pokemon affected by Power / Guard / Speed Swap have their stats reverted to base when switching out? Or is that an unnecessary concern?

xsn34kzx commented:
> It would be good to have tests checking that Pokemon affected by Power / Guard / Speed Swap have their stats reverted to base when switching out? Or is that an unnecessary concern?

@frutescens This should be fine since, on `SwitchSummonPhase`, we reset all the data in `summonData` where the temporary stats are saved. I also think Baton doesn't transfer anything besides stat stages and tags, so it should be good there too. I don't think this is a unnecessary concern though, we might need a general `summonData` test for other things that use it (Protean & Libero iirc, Transform, etc.)

frutescens commented:
Do you think that test is within the scope of this PR or not? Thank you for your explanation. I just know that switch ins/outs are weak points of PokeRogue's code in general so I thought it would be good to check.

xsn34kzx commented:
> Do you think that test is within the scope of this PR or not? Thank you for your explanation. I just know that switch ins/outs are weak points of PokeRogue's code in general so I thought it would be good to check.

Mirroring my response from Discord just so the discussion doesn't get lost:

It's probably out of scope just for the sake of not limiting a core game engine test for something like switch-outs (`SwitchOutPhase`) to this PR specifically, especially since I'd want something that crucial to be comprehensive (I'm not too savvy on what hinges on switch-outs as of right now or prepared for what bugs might come out of testing it now).

frutescens reviewed (approved):
- Tested in Endless Spliced --> Seems good
- Tested in Classic --> Also seems good.

ysh4296 commented on src/enums/stat.ts:
```diff
@@ -1,8 +1,75 @@
+}
+
+/** A constant array comprised of the {@linkcode Stat} values that make up {@linkcode PermanentStat}. */
+export const PERMANENT_STATS = [ Stat.HP, Stat.ATK, Stat.DEF, Stat.SPATK, Stat.SPDEF, Stat.SPD ] as const;
```
wondering if it's appropriate for enum files to include const values and functions

xsn34kzx replied:
Yeah, I more or less left everything in the current file just to get things working and to wait if the lead devs or anyone else wanted the additional types, constants, and functions in another companion file.

ysh4296 reviewed (approved):
It's impressive how you clearly separated the ambiguous parts of the Pokémon stat system and unified and adjusted the logic in the existing codebase.

As you mentioned in description, it seems that #2658  will need significant changes to align with the new stat system. If it's okay with you, I'd like to continue working on it and request review afterward.

xsn34kzx commented:
> It's impressive how you clearly separated the ambiguous parts of the Pokémon stat system and unified and adjusted the logic in the existing codebase.
> 
> As you mentioned in description, it seems that #2658 will need significant changes to align with the new stat system. If it's okay with you, I'd like to continue working on it and request review afterward.

I was actually gonna contact you when this got closer to getting into `beta`, it would definitely be welcomed if you still wanted to work on it! If you have any questions as well when making changes, feel free to ask.

NicusPulcis commented on src/locales/it/move-trigger.json:
```diff
@@ -3,6 +3,10 @@
   "absorbedElectricity": "{{pokemonName}} assorbe elettricità!",
   "switchedStatChanges": "{{pokemonName}} scambia con il bersaglio le modifiche alle statistiche!",
+  "switchedTwoStatChanges": "{{pokemonName}} scambia con il bersaglio le modifiche a {{firstStat}} e {{secondStat}}!",
+  "switchedStat": "{{pokemonName}} scambia la sua {{stat}} con quella del bersaglio!",
```
```suggestion
  "switchedStat": "{{pokemonName}} scambia il/la suo/a {{stat}} con quello/a del bersaglio!",
```

xsn34kzx replied:
In this specific case, the only move that uses this is Speed Swap, so I think you can use whatever article "Speed" uses for now.

NicusPulcis reviewed (commented):
added a suggestion regarding the italian localization due to stat names being male/female gendered in italian - hence needing both the male/female articles

xsn34kzx commented:
Fixed a few ability application functions that accidentally had their `simulate` arguments removed, which helped in addressing a bug reported on Discord: [White Smoke And Repeated Messages / double activation](https://discord.com/channels/1125469663833370665/1267263370730668143) (courtesy of @Snailman11)

https://github.com/user-attachments/assets/03e5e41a-c7e7-408d-b038-6f4e14fe7a22

https://github.com/user-attachments/assets/2d8a6d62-58ae-43fb-9062-a968905453e0

I also added a `TODO` for when tag applications can also be simulated (since this fix basically "silences" the message by making subsequent calls simulated, which can't be done for Mist's application message also being repeated).

NicusPulcis commented on src/locales/it/modifier-type.json:
```diff
@@ -49,8 +49,8 @@
-    "TempBattleStatBoosterModifierType": {
-      "description": "Aumenta {{tempBattleStatName}} di un livello a tutti i Pokémon nel gruppo per 5 battaglie."
+    "TempStatStageBoosterModifierType": {
+      "description": "Aumenta {{stat}} di un livello a tutti i Pokémon nel gruppo per 5 battaglie."
```
```suggestion
      "description": "Aumenta la statistica {{stat}} di un livello\na tutti i Pokémon nel gruppo per 5 battaglie."
```

NicusPulcis commented on src/locales/it/modifier-type.json:
```diff
@@ -61,8 +61,8 @@
-    "PokemonBaseStatBoosterModifierType": {
-      "description": "Aumenta {{statName}} di base del possessore del 10%."
+    "BaseStatBoosterModifierType": {
+      "description": "Aumenta {{stat}} di base del possessore del 10%."
```
```suggestion
      "description": "Aumenta l'/la {{stat}} di base del possessore del 10%."
```

NicusPulcis commented on src/locales/it/modifier-type.json:
```diff
@@ -61,8 +61,8 @@
     "AllPokemonLevelIncrementModifierType": {
       "description": "Aumenta i livell di tutti i Pokémon della squadra di {{levels}}."
```
```suggestion
      "description": "Aumenta il livello di tutti i Pokémon della squadra di {{levels}}."
```

NicusPulcis reviewed (approved):
italian localization and fixings done (there was a typo alr existing before the PR and i fixed that too)
